### PR TITLE
add deeplabv3P segmentation control for controlnet

### PR DESCRIPTION
**Pull request by Gaohan123**

### CODE REVIEW

**Comment by SamitHuang** [comment]
Please complete and upload the readme document asap.

**Comment by Gaohan123** [comment]
> Please complete and upload the readme document asap.

**Comment by Gaohan123** [comment]
> Please complete and upload the readme document asap.

Already complete

File: examples/stable_diffusion_v2/annotator/segmentation/__init__.py
Comment (SamitHuang) [inline examples/stable_diffusion_v2/annotator/segmentation/__init__.py:1]: align the name to conditions

File: examples/stable_diffusion_v2/controlnet.md
[the remark this answers is not in the document]
Comment (Gaohan123) [reply examples/stable_diffusion_v2/controlnet.md:1]: fixed

File: examples/stable_diffusion_v2/README.md [hunk 1]
```diff
@@ -39,7 +39,8 @@ For a quick tour, please view [demo](demo.md).
     - [Chinese Text-guided Image Inpainting](#chinese-text-guided-image-inpainting)
   - [Training](#training-1)
 - [Stable Diffusion with ControlNet](#stable-diffusion-with-controlnet)
-- [Stable Diffusion with T2I-Adapter](#stable-diffusion-with-t2i-adapter)
+  - [Inference](#inference-2)
+- [Stable Diffusion with T2I Adapter](#stable-diffusion-with-t2i-adapter)
```
Comment (hadipash) [inline examples/stable_diffusion_v2/README.md:43]: T2I-Adapter. I missed dash at line 332.

File: examples/stable_diffusion_v2/README.md [hunk 2]
```diff
@@ -324,6 +325,10 @@ after setting `data_path` in `run_train_v1.sh` to your dataset path.
 
 # Stable Diffusion with ControlNet
 
+## Inference
```
Comment (hadipash) [inline examples/stable_diffusion_v2/README.md:328]: Add a brief intro to ControlNet?

File: examples/stable_diffusion_v2/requirements.txt [hunk 3]
```diff
@@ -9,3 +9,7 @@ imagesize
 toolz
 pillow==9.1.1
 tqdm
+imageio==2.31.1
+webdataset==0.2.48
+addict==2.4.0
+yapf==0.40.1
```
Comment (hadipash) [inline examples/stable_diffusion_v2/requirements.txt:15]: don't add dependencies that are not used in the project.

File: examples/stable_diffusion_v2/image2image.py [hunk 4]
```diff
@@ -9,10 +9,10 @@
 
 import cv2
 import numpy as np
-from annotator.canny import CannyDetector
-from annotator.util import HWC3, resize_image
 from cldm.ddim_hacked import DDIMSampler
 from cldm.model import create_model, load_model
+from conditions.controlnet import CannyDetector, SegmentDetector
+from conditions.controlnet.utils.utils import HWC3, resize_image
```
Comment (hadipash) [inline examples/stable_diffusion_v2/image2image.py:15]: define `HWC3`, `resize_image` inside `utils.__init__.py` and you can avoid `from utils.utils`.

File: examples/stable_diffusion_v2/image2image.py [hunk 5]
```diff
@@ -120,16 +121,24 @@ def main(args):
...
-        )
-
-        control = ms.Tensor(detected_map.copy()).float() / 255.0
-        control = control.permute(2, 0, 1)
-        control = ops.stack([control for _ in range(num_samples)], axis=0)
+    elif args.mode == MODE["segmentation"]:
+        if os.path.exists(args.condition_ckpt_path):
+            apply_segment = SegmentDetector(ckpt_path=args.condition_ckpt_path)
+        else:
+            logger.warning(f"!!!Warning!!!: {args.condition_ckpt_path} doesn't exist")
```
Comment (hadipash) [inline examples/stable_diffusion_v2/image2image.py:128]:
```suggestion
            logger.warning(f"Segmentation Detector checkpoint {args.condition_ckpt_path} was not found.")
```
Please use more descriptive warnings for easier debugging.

File: examples/stable_diffusion_v2/image2image.py [hunk 6]
```diff
@@ -215,14 +229,17 @@ def decode_and_save_result(samples, detected_map, outpath, filename):
         "--mode",
         type=str,
         default="canny",
-        choices=[MODE["canny"]],
+        choices=[MODE["canny"], MODE["segmentation"]],
```
Comment (hadipash) [inline examples/stable_diffusion_v2/image2image.py:232]: `list(MODE.keys())`

File: examples/stable_diffusion_v2/README.md (hunk 1 above)
Comment (Gaohan123) [reply examples/stable_diffusion_v2/README.md]: fixed

File: examples/stable_diffusion_v2/README.md (hunk 2 above)
Comment (Gaohan123) [reply examples/stable_diffusion_v2/README.md:330]: fixed

File: examples/stable_diffusion_v2/requirements.txt (hunk 3 above)
Comment (Gaohan123) [reply examples/stable_diffusion_v2/requirements.txt]: fixed

File: examples/stable_diffusion_v2/image2image.py (hunk 5 above)
Comment (Gaohan123) [reply examples/stable_diffusion_v2/image2image.py]: fixed

File: examples/stable_diffusion_v2/image2image.py (hunk 6 above)
Comment (Gaohan123) [reply examples/stable_diffusion_v2/image2image.py]: fixed

**Comment by congw729** [comment]
LGTM

**Review by congw729 — APPROVED** [review]

**Review by hadipash — APPROVED** [review]

File: examples/stable_diffusion_v2/image2image.py
Comment (SamitHuang) [inline examples/stable_diffusion_v2/image2image.py:1]: Similar to t2i adapter, rename this script name to `controlnet_image2image.py`. 
`image2image.py` refers to the original SD image-to-image pipeline.

File: examples/stable_diffusion_v2/conditions/controlnet/canny_detector.py
Comment (SamitHuang) [inline examples/stable_diffusion_v2/conditions/controlnet/canny_detector.py:1]: the subfolders of `conditions` are group by  control signal types. Please follow the file structure.
```test
conditions\
   depth\
      
   canny\
   segmentation\
   motion\
```

File: examples/stable_diffusion_v2/README.md [hunk 7]
```diff
@@ -324,7 +325,13 @@ after setting `data_path` in `run_train_v1.sh` to your dataset path.
 
 # Stable Diffusion with ControlNet
 
-# Stable Diffusion with T2I Adapter
+ControlNet controls pretrained large diffusion models to support additional input conditions. The ControlNet learns task-specific conditions in an end-to-end way, and the learning is robust even when the training dataset is small. Large diffusion models like Stable Diffusion can be augmented with ControlNets to enable conditional inputs like canny edge maps, segmentation maps, keypoints, etc.
```
Comment (SamitHuang) [inline examples/stable_diffusion_v2/README.md:328]: Please add hyper-link to the controlnet readme

File: examples/stable_diffusion_v2/image2image.py
Comment (Gaohan123) [reply examples/stable_diffusion_v2/image2image.py:1]: fixed

File: examples/stable_diffusion_v2/conditions/controlnet/canny_detector.py
Comment (Gaohan123) [reply examples/stable_diffusion_v2/conditions/controlnet/canny_detector.py:1]: fixed

File: examples/stable_diffusion_v2/README.md (hunk 7 above)
Comment (Gaohan123) [reply examples/stable_diffusion_v2/README.md:328]: It already has in line 332

**Review by SamitHuang — APPROVED** [review]

**Pull request merged**